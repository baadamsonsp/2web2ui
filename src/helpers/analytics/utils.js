import _ from 'lodash';
import { ANALYTICS_FORM_EVENT_CATEGORY, ANALYTICS_WHITELISTED_FORMS } from '../../constants';

export function isWhitelistedForm(event) {
  const formName = _.first(event.label.split(':'));
  return event.category === ANALYTICS_FORM_EVENT_CATEGORY && _.includes(ANALYTICS_WHITELISTED_FORMS, formName);
}

export function determineFormValidationState(action) {
  const fieldErrors = _.map(action.payload.syncErrors, (val, key) => `${key}=${_.isString(val) ? val.replace('=','%3D') : JSON.stringify(val)}`);

  return {
    action: `Validation ${fieldErrors.length ? 'Error' : 'Success'}`,
    label: `${action.meta.form}: ${fieldErrors.join(',')}`
  };
}
