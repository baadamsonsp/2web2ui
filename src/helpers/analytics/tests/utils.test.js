import cases from 'jest-in-case';
import * as utils from '../utils';

describe('Analytics Utils', () => {
  const isWhitelistedFormTestCases = {
    'true when category is Form, form name is whitelisted one': {
      event: { category: 'Form', label: 'loginForm' },
      expectation: true
    },
    'true when category is Form, form name is whitelisted one with validation errors': {
      event: {
        category: 'Form',
        label: 'loginForm: name=Required,email=Required'
      }, expectation: true
    },
    'false when category is Form, form name is not whitelisted one': {
      event: { category: 'Form', label: 'apiKeyForm' },
      expectation: false
    },
    'false when category is not Form, form name is whitelisted one': {
      event: { category: 'Foo', label: 'loginForm' },
      expectation: false
    }
  };

  cases('isWhitelistedForm', ({ name, event, expectation }) => {
    expect(utils.isWhitelistedForm(event)).toBe(expectation);
  }, isWhitelistedFormTestCases);

  const determineFormValidationStateTestCases = {
    'validation success with no syncErrors': { action: { payload: { syncErrors: {}}, meta: { form: 'fooForm' }}},
    'validation error with fields errors': {
      action: {
        payload: { syncErrors: { name: 'Required' }},
        meta: { form: 'fooForm' }
      }
    },
    'validation error with generic error': {
      action: {
        payload: { syncErrors: { _error: 'Some form level error' }},
        meta: { form: 'fooForm' }
      }
    },
    'validation error with generic error with equal sign': {
      action: {
        payload: { syncErrors: { _error: 'must be value=2' }},
        meta: { form: 'fooForm' }
      }
    },
    'serializes when syncErrors contains object in value': {
      action: {
        payload: { syncErrors: { billingInfo: { address: 'Required', zip: 'Required' }}},
        meta: { form: 'fooForm' }
      }
    }
  };

  cases('determineFormValidationState', ({ name, action }) => {
    expect(utils.determineFormValidationState(action)).toMatchSnapshot();
  }, determineFormValidationStateTestCases);


});
