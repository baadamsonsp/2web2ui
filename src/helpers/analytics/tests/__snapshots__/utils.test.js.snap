// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics Utils determineFormValidationState serializes when syncErrors contains object in value 1`] = `
Object {
  "action": "Validation Error",
  "label": "fooForm: billingInfo={\\"address\\":\\"Required\\",\\"zip\\":\\"Required\\"}",
}
`;

exports[`Analytics Utils determineFormValidationState validation error with fields errors 1`] = `
Object {
  "action": "Validation Error",
  "label": "fooForm: name=Required",
}
`;

exports[`Analytics Utils determineFormValidationState validation error with generic error 1`] = `
Object {
  "action": "Validation Error",
  "label": "fooForm: _error=Some form level error",
}
`;

exports[`Analytics Utils determineFormValidationState validation error with generic error with equal sign 1`] = `
Object {
  "action": "Validation Error",
  "label": "fooForm: _error=must be value%3D2",
}
`;

exports[`Analytics Utils determineFormValidationState validation success with no syncErrors 1`] = `
Object {
  "action": "Validation Success",
  "label": "fooForm: ",
}
`;
